fix: Replace generic AI connection error with detailed API response a…

syuzanvs opened the pull request:
fix: Replace generic AI connection error with detailed API response and fallback status messages

- Improved error response handling by ensuring that the error data is parsed and logged correctly when the response is not OK.
- Added more descriptive error messages for various HTTP status codes (401, 404, 500) to provide clearer feedback to users.
- Refactored the JSON validation logic to ensure robust error handling when parsing error messages.

Closes #55

syuzanvs committed:
fix: Replace generic AI connection error with detailed API response and fallback status messages

- Improved error response handling by ensuring that the error data is parsed and logged correctly when the response is not OK.
- Added more descriptive error messages for various HTTP status codes (401, 404, 500) to provide clearer feedback to users.
- Refactored the JSON validation logic to ensure robust error handling when parsing error messages.

Closes #55

jjroelofs reviewed (changes requested):
Thanks, but there are a lot of status codes missing here, some of which are common and important in the (Open)AI ecosystem. Please either use a library or copy the list of codes and messages from a library. Whichever is easier to integrate with the CKEditor 5 translation layer:

* https://github.com/prettymuchbryce/http-status-codes
* https://github.com/prettymuchbryce/http-status-codes/blob/master/codes.json

We want to support all status codes in the above source.

jjroelofs commented on src/aiagentservice.js:
```diff
@@ -308,13 +313,28 @@ export default class AiAgentService {
                 case 'ReadableStream not supported':
                     errorMessage = t('Browser does not support readable streams');
                     break;
                 case 'AiAgent: Fetch failed':
```
Also, please check if this case is not preventing more precise error messages from materializing. Maybe only keep this message in the default behavior? I'm not sure I didn't do a deep analysis here but please analyze this carefully

syuzanvs committed:
Fix conflicts

syuzanvs committed:
fix: Replace generic AI connection error with detailed API response and fallback status messages

Improve error handling in fetchAndProcessGptResponse method by adding specific user-friendly messages for 403 (Forbidden), 429 (Too Many Requests), and 503 (Service Unavailable) status codes. This enhancement provides clearer feedback to users regarding API access and server issues.

Closes #55

jjroelofs reviewed (changes requested):
1. There are still 48 status codes missing, so the biggest problem with the PR is exactly the same problem as the last release cycle
2. I don't want to dump all 58 status code messages to aiagentservice.ts, either put them in a utility or service file, or just add the dependency to our project and integrate the messages with our translation layer some way.
3. Now that I'm thinking of it, our translation files need to be updated with translation strings for all english strings in our project, that goes for these too. You can just use Google Translate or ChatGPT for this: 
https://github.com/dxpr/ckeditor5-ai-agent/tree/1.x/lang

Don't spend too much time on doing this, just timebox half an hour to migrate the strings to the contexts.json and add the translations with ChatGPT or Google Translate.

syuzanvs committed:
fix: Replace generic AI connection error with detailed API response and fallback status messages

Refactor error handling in fetchAndProcessGptResponse method

- Improved error message handling for AI connection issues.
- Enhanced JSON validation for error messages.
- Added retry logic for API calls with remaining attempts.
- Ensured user-friendly error messages are displayed.
- Added translations for unexpected error messages.

Closes #55

jjroelofs reviewed (changes requested):
Thanks, great work very clean now.
It seems that the JSON parsing is getting in the way of relaying error messages. When I set CKEditor config to `endpointUrl: 'https://httpstat.us/402'` I expect to see the 402 error message but instead I get:

`Unexpected non-whitespace character after JSON at position 4 (line 1 column 5)`

I understand we need to parse JSON to check for the detailed message, but many servers do not return JSON when there is an error, so we have to account for this situation too.

syuzanvs committed:
fix: Replace generic AI connection error with detailed API response and fallback status messages

Improve error handling in getError method

- Enhanced the getError method to handle different content types more robustly.
- Added checks for response content types to ensure proper error message extraction.
- Ensured that error messages are consistently formatted as JSON strings for better error reporting.

Closes #55

jjroelofs reviewed (changes requested):
## Translation bug

Thanks, great work. Everything seems to be working perfectly now.

Except for translation.

On this branch run `yarn run dll:build;yarn run dll:serve`

You can confirm Spanish translation is working by looking at the placeholder text:

<img width="832" alt="Screenshot 2024-12-23 at 20 12 24" src="https://github.com/user-attachments/assets/d30fe2b5-82f0-4bde-bef6-b488a893109e" />

Consequently we expect the error message to be Spanish but it is actually in English:

<img width="822" alt="Screenshot 2024-12-23 at 20 13 05" src="https://github.com/user-attachments/assets/3c2ef392-68bc-4e67-9319-8d14d8e2f108" />

It does appear we account for this string in our translation file so something must be wrong in our implementation of the translation layer:


<img width="525" alt="Screenshot 2024-12-23 at 20 12 48" src="https://github.com/user-attachments/assets/741ad345-b6a6-44ef-a361-585e26931633" />

## Translation extra work

If you pull 1.x the DLL built will default to arabic language. Please also add arabic translations and see if the notification is looking OK. I don't think its a big deal it appears on the right-hand side, but bonus points if you move it to the other side when detecting RTL language (follow CKEditor localization guidelines for this though).

syuzanvs committed:
Merge branch '1.x' of github.com:dxpr/ckeditor5-ai-agent into syuzan/1.x/#55-replace-generic-ai-connection-error-with-detailed-api-response-and-fallback-status-messages

syuzanvs committed:
fix: Replace generic AI connection error with detailed API response and fallback status messages

- Updated `getErrorMessages` in `translations.ts` to improve error message retrieval.
- Refactored `AiAgentUI` to ensure better UI component initialization and error handling.
- Enhanced `AiAgentService` to improve moderation checks and error handling during API calls.
- Updated constants in `const.ts` to include new AI model configurations.
- Improved documentation and comments for clarity and maintainability.

Closes #55
